### PR TITLE
feat: [SSCA-1151]: support k8s container image match for artifact listing

### PR DESCRIPTION
**Pull request by inderschera**
## Describe your changes
- Added feature flag for changes.
- Made sure our changes work with both map and list of container list.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/57143)

### CODE REVIEW

**Comment by inderschera** [comment]
trigger uts

**Comment by inderschera** [comment]
trigger commonchecks

**Comment by inderschera** [comment]
trigger uts

**Comment by inderschera** [comment]
trigger uts

**Review by deepharness — CHANGES_REQUESTED** [review]
> Reviewed 5 of 12 files at r1, 1 of 4 files at r2, 1 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/57143)* status: 8 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`ssca-manager/service/src/main/java/io/harness/changestreams/eventhandlers/InstanceNGRedisEventHandler.java` line 38 at r3](https://reviewable.io/reviews/harness/harness-core/57143#-NmjEA7q0uNVGZLRffNL:-NmjEA7q0uNVGZLRffNM:b-4nueh) ([raw file](https://github.com/harness/harness-core/blob/ff6f5fe558ec6fbf9eafa8dec3a4adcdfbd7ad0a/ssca-manager/service/src/main/java/io/harness/changestreams/eventhandlers/InstanceNGRedisEventHandler.java#L38)):*
> ```Java
>     if (featureFlagService.isFeatureFlagEnabled(
>             instanceDTO.getAccountIdentifier(), FeatureName.SSCA_MATCH_INSTANCE_IMAGE_NAME.name())) {
>       if (instanceInfo != null && K8S_INSTANCE_INFO_CLASS.equals(instanceInfo.get(_CLASS))) {
> ```

you can club it condition , 2 if blocks are not required
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java` line 82 at r3](https://reviewable.io/reviews/harness/harness-core/57143#-NmjImlODG6ZW9QzD31S:-NmjImlP-7CWDwDm4yjA:b-4gpyfu) ([raw file](https://github.com/harness/harness-core/blob/ff6f5fe558ec6fbf9eafa8dec3a4adcdfbd7ad0a/ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java#L82)):*
> ```Java
>       if (featureFlagService.isFeatureFlagEnabled(
>               instance.getAccountIdentifier(), FeatureName.SSCA_MATCH_INSTANCE_IMAGE_NAME.name())) {
>         if (cdInstanceSummaryServiceHelper.isK8sInstanceInfo(instance)) {
> ```

avoid nesting 
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java` line 144 at r3](https://reviewable.io/reviews/harness/harness-core/57143#-NmjMdnB9qSUwJvsGtTP:-NmjMdnB9qSUwJvsGtTQ:b3lgh92) ([raw file](https://github.com/harness/harness-core/blob/ff6f5fe558ec6fbf9eafa8dec3a4adcdfbd7ad0a/ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java#L144)):*
> ```Java
>       if (featureFlagService.isFeatureFlagEnabled(
>               instance.getAccountIdentifier(), FeatureName.SSCA_MATCH_INSTANCE_IMAGE_NAME.name())) {
>         if (cdInstanceSummaryServiceHelper.isK8sInstanceInfo(instance)) {
> ```

avoid nested if
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java` line 177 at r3](https://reviewable.io/reviews/harness/harness-core/57143#-NmjNH0Z0kHCesckKxy3:-NmjNH0Z0kHCesckKxy4:b-mlr6ik) ([raw file](https://github.com/harness/harness-core/blob/ff6f5fe558ec6fbf9eafa8dec3a4adcdfbd7ad0a/ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java#L177)):*
> ```Java
>       if (cdInstanceSummary.getInstanceIds().isEmpty()) {
>         artifactService.updateArtifactEnvCount(artifact, cdInstanceSummary.getEnvType(), -1);
>         cdInstanceSummaryRepo.delete(cdInstanceSummary);
> ```

Will it be logged in repo ? if not we should log this

**Review by inderschera — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/57143)* status: 8 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`ssca-manager/service/src/main/java/io/harness/changestreams/eventhandlers/InstanceNGRedisEventHandler.java` line 38 at r3](https://reviewable.io/reviews/harness/harness-core/57143#-NmjEA7q0uNVGZLRffNL:-NmjQ_B5EHsZFv2ShbW3:b4ll428) ([raw file](https://github.com/harness/harness-core/blob/ff6f5fe558ec6fbf9eafa8dec3a4adcdfbd7ad0a/ssca-manager/service/src/main/java/io/harness/changestreams/eventhandlers/InstanceNGRedisEventHandler.java#L38)):*
<details><summary><i>Previously, deepharness wrote…</i></summary><blockquote>

you can club it condition , 2 if blocks are not required
</blockquote></details>

keeping this for future in case we have to do the same for other instance info.
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java` line 82 at r3](https://reviewable.io/reviews/harness/harness-core/57143#-NmjImlODG6ZW9QzD31S:-NmjQhDC07-KxHhZ3fDk:b-4jna3w) ([raw file](https://github.com/harness/harness-core/blob/ff6f5fe558ec6fbf9eafa8dec3a4adcdfbd7ad0a/ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java#L82)):*
<details><summary><i>Previously, deepharness wrote…</i></summary><blockquote>

avoid nesting
</blockquote></details>

added this for logging in else condition.
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java` line 144 at r3](https://reviewable.io/reviews/harness/harness-core/57143#-NmjMdnB9qSUwJvsGtTP:-NmjQktG8-DEP22fmWsQ:b-g6ow4m) ([raw file](https://github.com/harness/harness-core/blob/ff6f5fe558ec6fbf9eafa8dec3a4adcdfbd7ad0a/ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java#L144)):*
<details><summary><i>Previously, deepharness wrote…</i></summary><blockquote>

avoid nested if
</blockquote></details>

added this for logging in else condition
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java` line 177 at r3](https://reviewable.io/reviews/harness/harness-core/57143#-NmjNH0Z0kHCesckKxy3:-NmjQlZj8RX4g6Raz4qj:b6n07oh) ([raw file](https://github.com/harness/harness-core/blob/ff6f5fe558ec6fbf9eafa8dec3a4adcdfbd7ad0a/ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java#L177)):*
<details><summary><i>Previously, deepharness wrote…</i></summary><blockquote>

Will it be logged in repo ? if not we should log this
</blockquote></details>

didn't understand this

**Review by deepharness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/57143)* status: 8 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`ssca-manager/service/src/main/java/io/harness/changestreams/eventhandlers/InstanceNGRedisEventHandler.java` line 38 at r3](https://reviewable.io/reviews/harness/harness-core/57143#-NmjEA7q0uNVGZLRffNL:-NmjRBTO4fQV-pAsVktr:bq6vhzg) ([raw file](https://github.com/harness/harness-core/blob/ff6f5fe558ec6fbf9eafa8dec3a4adcdfbd7ad0a/ssca-manager/service/src/main/java/io/harness/changestreams/eventhandlers/InstanceNGRedisEventHandler.java#L38)):*
<details><summary><i>Previously, inderschera (Inderpreet Chera) wrote…</i></summary><blockquote>

keeping this for future in case we have to do the same for other instance info.
</blockquote></details>

follow YAGNI
https://martinfowler.com/bliki/Yagni.html
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java` line 82 at r3](https://reviewable.io/reviews/harness/harness-core/57143#-NmjImlODG6ZW9QzD31S:-NmjRfFeCjKSsMId_egd:bb1jmco) ([raw file](https://github.com/harness/harness-core/blob/ff6f5fe558ec6fbf9eafa8dec3a4adcdfbd7ad0a/ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java#L82)):*
<details><summary><i>Previously, inderschera (Inderpreet Chera) wrote…</i></summary><blockquote>

added this for logging in else condition.
</blockquote></details>

you can inverse the condition then instead of nesting
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java` line 144 at r3](https://reviewable.io/reviews/harness/harness-core/57143#-NmjMdnB9qSUwJvsGtTP:-NmjRstU6w2zmLYqFi14:bb1jmco) ([raw file](https://github.com/harness/harness-core/blob/ff6f5fe558ec6fbf9eafa8dec3a4adcdfbd7ad0a/ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java#L144)):*
<details><summary><i>Previously, inderschera (Inderpreet Chera) wrote…</i></summary><blockquote>

added this for logging in else condition
</blockquote></details>

you can inverse the condition then instead of nesting
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java` line 177 at r3](https://reviewable.io/reviews/harness/harness-core/57143#-NmjNH0Z0kHCesckKxy3:-NmjRuWa8OIWvowvm3DE:b-dx2517) ([raw file](https://github.com/harness/harness-core/blob/ff6f5fe558ec6fbf9eafa8dec3a4adcdfbd7ad0a/ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java#L177)):*
<details><summary><i>Previously, inderschera (Inderpreet Chera) wrote…</i></summary><blockquote>

didn't understand this
</blockquote></details>

are we logging this , if cd instance is deleted ? As I can see there are many conditional flows , instance collection may not be deleted . We should get these info in logs so that its easier to debug than checking it in db

**Review by inderschera — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/57143)* status: 8 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java` line 177 at r3](https://reviewable.io/reviews/harness/harness-core/57143#-NmjNH0Z0kHCesckKxy3:-NmjXpr701t8USWV4K8X:b-l0zuod) ([raw file](https://github.com/harness/harness-core/blob/ff6f5fe558ec6fbf9eafa8dec3a4adcdfbd7ad0a/ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java#L177)):*
<details><summary><i>Previously, deepharness wrote…</i></summary><blockquote>

are we logging this , if cd instance is deleted ? As I can see there are many conditional flows , instance collection may not be deleted . We should get these info in logs so that its easier to debug than checking it in db
</blockquote></details>

We have logs everywhere when it's not deleted.

**Review by inderschera — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/57143)* status: 9 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`ssca-manager/service/src/main/java/io/harness/changestreams/eventhandlers/InstanceNGRedisEventHandler.java` line 38 at r3](https://reviewable.io/reviews/harness/harness-core/57143#-NmjEA7q0uNVGZLRffNL:-Nmjajta5IFX7h84WIMK:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/ff6f5fe558ec6fbf9eafa8dec3a4adcdfbd7ad0a/ssca-manager/service/src/main/java/io/harness/changestreams/eventhandlers/InstanceNGRedisEventHandler.java#L38)):*
<details><summary><i>Previously, deepharness wrote…</i></summary><blockquote>

follow YAGNI
https://martinfowler.com/bliki/Yagni.html
</blockquote></details>

Done.
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java` line 177 at r3](https://reviewable.io/reviews/harness/harness-core/57143#-NmjNH0Z0kHCesckKxy3:-NmjaxHk4KNGsBEi8gXH:b-7rl0kd) ([raw file](https://github.com/harness/harness-core/blob/ff6f5fe558ec6fbf9eafa8dec3a4adcdfbd7ad0a/ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java#L177)):*
<details><summary><i>Previously, inderschera (Inderpreet Chera) wrote…</i></summary><blockquote>

We have logs everywhere when it's not deleted.
</blockquote></details>

done

**Review by deepharness — APPROVED** [review]
> Reviewed 4 of 12 files at r1, 2 of 4 files at r2, 1 of 2 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/57143)* status: <span>:shipit:</span> complete! ✓

**Comment by Varsha-lalwani** [comment]
trigger uts

File: ssca-manager/service/src/test/resources/k8s-instance-with-container-list.json [hunk 1]
```diff
@@ -0,0 +1,52 @@
+{
```
Comment (reetika04) [inline ssca-manager/service/src/test/resources/k8s-instance-with-container-list.json:1]: Given that this a test resource, please use dummy values. Do not use any real values.

**Comment by reetika04** [comment]
trigger uts

**Pull request merged**